File: scripts/rename-by-qr.py
#!/usr/bin/env python3
"""
rename-by-qr.py

Reads all PNG/JPG images in a folder, decodes the QR code in each,
extracts the certificate/card ID from the URL, and renames the file.

Usage:
  python3 scripts/rename-by-qr.py <path-to-folder>

Examples:
  python3 scripts/rename-by-qr.py ./recovered_records/certificates
  python3 scripts/rename-by-qr.py ./recovered_records/cards
"""

import sys
import os
import re
from PIL import Image
from pyzbar.pyzbar import decode

def extract_id_from_url(url: str):
    """Extract the record ID from the QR code URL."""
    patterns = [
        r'/certificate/view/(certificate_\d+)',
        r'/card/view/(c-\d+)',
        r'/operator/view/([^\s/]+)',
        r'/aasia-steel-card/view/([^\s/]+)',
    ]
    for pattern in patterns:
        m = re.search(pattern, url)
        if m:
            return m.group(1)
    return None

def process_file(folder_path: str, filename: str):
    file_path = os.path.join(folder_path, filename)
    ext = os.path.splitext(filename)[1]

    try:
        img = Image.open(file_path).convert("RGB")

        # Try 1: Decode as-is
        results = decode(img)

        # Try 2: Scale up if too small
        if not results:
            w, h = img.size
            MIN_DIM = 1200
            if w < MIN_DIM or h < MIN_DIM:
                scale = MIN_DIM / min(w, h)
                img_up = img.resize((int(w * scale), int(h * scale)), Image.LANCZOS)
                results = decode(img_up)

        if not results:
            print(f"⚠️  [SKIP] No QR code found in: {filename}")
            return "skip"

        qr_url = results[0].data.decode("utf-8")
        print(f"    QR URL: {qr_url}")

        new_id = extract_id_from_url(qr_url)
        if not new_id:
            print(f"⚠️  [SKIP] Could not extract ID from QR URL in: {filename}")
            print(f"           URL was: {qr_url}")
            return "skip"

        new_filename = f"{new_id}{ext}"
        new_path = os.path.join(folder_path, new_filename)

        if os.path.exists(new_path) and new_path != file_path:
            print(f"⚠️  [SKIP] Target already exists, skipping: {new_filename}")
            return "skip"

        if new_filename == filename:
            print(f"✅  [OK]   Already correctly named: {filename}")
            return "ok"

        os.rename(file_path, new_path)
        print(f"✅  [RENAMED] {filename}  →  {new_filename}")
        return "ok"

    except Exception as e:
        print(f"❌  [ERROR] Failed to process: {filename}")
        print(f"           {e}")
        return "error"


def main():
    if len(sys.argv) < 2:
        print("❌  Please provide a folder path.")
        print("    Usage: python3 scripts/rename-by-qr.py <folder>")
        sys.exit(1)

    folder_path = sys.argv[1]
    if not os.path.isdir(folder_path):
        print(f"❌  Folder not found: {folder_path}")
        sys.exit(1)

    files = sorted([
        f for f in os.listdir(folder_path)
        if f.lower().endswith((".png", ".jpg", ".jpeg"))
    ])

    if not files:
        print("⚠️  No image files found in the folder.")
        sys.exit(0)

    print(f"\n📂 Found {len(files)} image(s) in: {folder_path}\n")

    ok = skip = err = 0
    for f in files:
        print(f"🔍 Processing: {f}")
        result = process_file(folder_path, f)
        if result == "ok":
            ok += 1
        elif result == "skip":
            skip += 1
        else:
            err += 1

    print("\n─────────────────────────────────────")
    print(f"✅  Renamed : {ok}")
    print(f"⚠️  Skipped : {skip}")
    print(f"❌  Errors  : {err}")
    print("─────────────────────────────────────\n")


if __name__ == "__main__":
    main()

File: scripts/rename-by-qr.js
/**
 * rename-by-qr.js
 *
 * Reads all PNG images in a folder, decodes the QR code in each image,
 * extracts the certificate/card ID from the URL, and renames the file.
 *
 * Usage:
 *   node scripts/rename-by-qr.js <path-to-folder>
 *
 * Example:
 *   node scripts/rename-by-qr.js ./input/certificates
 *   node scripts/rename-by-qr.js ./input/cards
 */

const { Jimp } = require("jimp");
const jsQR = require("jsqr");
const fs = require("fs");
const path = require("path");

const folderPath = process.argv[2];

if (!folderPath) {
  console.error("❌  Please provide a folder path as an argument.");
  console.error("    Usage: node scripts/rename-by-qr.js <path-to-folder>");
  process.exit(1);
}

if (!fs.existsSync(folderPath)) {
  console.error(`❌  Folder not found: ${folderPath}`);
  process.exit(1);
}

// Get all PNG files in the folder
const files = fs.readdirSync(folderPath).filter((f) => f.match(/\.(png|jpg|jpeg)$/i));

if (files.length === 0) {
  console.log("⚠️  No image files found in the folder.");
  process.exit(0);
}

console.log(`\n📂 Found ${files.length} image(s) in: ${folderPath}\n`);

let successCount = 0;
let skipCount = 0;
let failCount = 0;

async function processFile(file) {
  const filePath = path.join(folderPath, file);

  try {
    const rawImage = await Jimp.read(filePath);

    // --- Attempt 1: Preprocess - greyscale + contrast boost + scale up ---
    const img1 = rawImage.clone();
    // Scale up small images so jsQR can see the QR module pixels better
    const MIN_DIM = 1200;
    if (img1.bitmap.width < MIN_DIM || img1.bitmap.height < MIN_DIM) {
      const scale = MIN_DIM / Math.min(img1.bitmap.width, img1.bitmap.height);
      img1.scale(scale);
    }
    img1.greyscale().contrast(0.8).normalize();

    let code = jsQR(
      new Uint8ClampedArray(img1.bitmap.data),
      img1.bitmap.width,
      img1.bitmap.height
    );

    // --- Attempt 2: Try inverted colors (some QR codes appear light-on-dark) ---
    if (!code) {
      const img2 = img1.clone().invert();
      code = jsQR(
        new Uint8ClampedArray(img2.bitmap.data),
        img2.bitmap.width,
        img2.bitmap.height
      );
    }

    // --- Attempt 3: Try original without preprocessing (raw) ---
    if (!code) {
      code = jsQR(
        new Uint8ClampedArray(rawImage.bitmap.data),
        rawImage.bitmap.width,
        rawImage.bitmap.height
      );
    }

    if (!code) {
      console.warn(`⚠️  [SKIP] No QR code found in: ${file}`);
      skipCount++;
      return;
    }

    const qrUrl = code.data;
    console.log(`    QR URL: ${qrUrl}`);

    // Try to extract certificate or card ID from the URL
    // Matches patterns like: /certificate/view/certificate_63 or /card/view/c-10
    const certMatch = qrUrl.match(/certificate\/view\/(certificate_\d+)/);
    const cardMatch = qrUrl.match(/card\/view\/(c-\d+)/);
    const operatorMatch = qrUrl.match(/operator\/view\/([^\s/]+)/);
    const aasiaMatch = qrUrl.match(/aasia-steel-card\/view\/([^\s/]+)/);

    let newId = null;
    if (certMatch) newId = certMatch[1];
    else if (cardMatch) newId = cardMatch[1];
    else if (operatorMatch) newId = operatorMatch[1];
    else if (aasiaMatch) newId = aasiaMatch[1];

    if (!newId) {
      console.warn(`⚠️  [SKIP] Could not extract ID from QR URL in: ${file}`);
      console.warn(`           URL was: ${qrUrl}`);
      skipCount++;
      return;
    }

    const ext = path.extname(file);
    const newFilename = `${newId}${ext}`;
    const newFilePath = path.join(folderPath, newFilename);

    if (fs.existsSync(newFilePath) && newFilePath !== filePath) {
      console.warn(`⚠️  [SKIP] Target file already exists, skipping: ${newFilename}`);
      skipCount++;
      return;
    }

    if (newFilename === file) {
      console.log(`✅  [OK]   Already correctly named: ${file}`);
      successCount++;
      return;
    }

    fs.renameSync(filePath, newFilePath);
    console.log(`✅  [RENAMED] ${file}  →  ${newFilename}`);
    successCount++;
  } catch (err) {
    console.error(`❌  [ERROR] Failed to process: ${file}`);
    console.error(`           ${err.message}`);
    failCount++;
  }
}

(async () => {
  for (const file of files) {
    process.stdout.write(`🔍 Processing: ${file}\n`);
    await processFile(file);
  }

  console.log("\n─────────────────────────────────────");
  console.log(`✅  Renamed : ${successCount}`);
  console.log(`⚠️  Skipped : ${skipCount}`);
  console.log(`❌  Errors  : ${failCount}`);
  console.log("─────────────────────────────────────\n");
})();
